### ISSUE TITLE
[Bug]: Controlnet models are loading from cache and predprocess after update each generation

### ISSUE THREAD

**Issue by FedorTarasow** [Mikubill/sd-webui-controlnet#2119]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits of both this extension and the webui

### What happened?

After updating to 1.6.0 controlnet works much slower than before. 

### Steps to reproduce the problem

1. Go to Img2Img inpaint mode
2. Activate any controlnet model
3. Generate Image.
4. Each time, I see in log "Loading model from cache", even If I have already generated images with this model. In addition I see in log "Loading preprocessor" despite I dont want to predprocess Image, it is already predprocessed.
![Screenshot_10](https://github.com/Mikubill/sd-webui-controlnet/assets/122746903/72dd0c9e-ee1c-4a2f-b965-bf8fdca5eb8f)

### What should have happened?

Models should load only once after first generation and if I click "Generate again it should work faster. Images should predprocess each generation instead they should use alredy preprocessed image

### Commit where the problem happens

webui: v1.6.0
controlnet: v1.1.410

### What browsers do you use to access the UI ?

Google Chrome

### Command Line Arguments

```Shell
No
```

### List of enabled extensions

![Screenshot_9](https://github.com/Mikubill/sd-webui-controlnet/assets/122746903/204b264c-c393-4fed-8175-411adc0a7f5e)

### Console logs

```Shell
venv "E:\stable-diffusion-webui\venv\Scripts\Python.exe"
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Version: v1.6.0
Commit hash: 5ef669de080814067961f28357256e8fe27544f4
Checking ReActor requirements... Ok
Launching Web UI with arguments:
no module 'xformers'. Processing without...
no module 'xformers'. Processing without...
No module 'xformers'. Proceeding without it.

========================= a1111-sd-webui-lycoris =========================
Starting from stable-diffusion-webui version 1.5.0
a1111-sd-webui-lycoris extension is no longer needed

All its features have been integrated into the native LoRA extension
LyCORIS models can now be used as if there are regular LoRA models

This extension has been automatically deactivated
Please remove this extension
==========================================================================

[-] ADetailer initialized. version: 23.9.2, num models: 9
2023-09-16 21:54:44,136 - ControlNet - INFO - ControlNet v1.1.410
ControlNet preprocessor location: E:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads
2023-09-16 21:54:44,289 - ControlNet - INFO - ControlNet v1.1.410
21:54:46 - ReActor - STATUS - Running v0.4.1
Loading weights [9aba26abdf] from E:\stable-diffusion-webui\models\Stable-diffusion\deliberate_v2.safetensors
Creating model from config: E:\stable-diffusion-webui\configs\v1-inference.yaml
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 29.8s (prepare environment: 8.2s, import torch: 3.7s, import gradio: 1.2s, setup paths: 1.0s, initialize shared: 0.3s, other imports: 0.9s, setup codeformer: 0.3s, load scripts: 12.3s, create ui: 1.2s, gradio launch: 0.5s).
Applying attention optimization: Doggettx... done.
Model loaded in 15.1s (load weights from disk: 1.1s, create model: 0.7s, apply weights to model: 10.8s, calculate empty prompt: 2.4s).
2023-09-16 21:57:44,513 - ControlNet - STATUS - Preview Resolution = 650
WARNING:fvcore.common.config:Loading config E:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\oneformer\configs/ade20k/oneformer_swin_large_IN21k_384_bs16_160k.yaml with yaml.unsafe_load. Your machine may be at risk if the file contains malicious content.
WARNING:fvcore.common.config:Loading config E:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\oneformer\configs/ade20k\Base-ADE20K-UnifiedSegmentation.yaml with yaml.unsafe_load. Your machine may be at risk if the file contains malicious content.
2023-09-16 21:58:19,509 - ControlNet - STATUS - Loading model: control_v11p_sd15_seg [e1f51eb9]
2023-09-16 21:58:22,596 - ControlNet - STATUS - Loaded state_dict from [E:\stable-diffusion-webui\models\ControlNet\control_v11p_sd15_seg.pth]
2023-09-16 21:58:22,596 - ControlNet - STATUS - controlnet_default_config
2023-09-16 21:58:24,962 - ControlNet - STATUS - ControlNet model control_v11p_sd15_seg [e1f51eb9] loaded.
2023-09-16 21:58:38,002 - ControlNet - STATUS - Loading preprocessor: oneformer_ade20k
2023-09-16 21:58:38,002 - ControlNet - STATUS - preprocessor resolution = 600
2023-09-16 21:58:38,814 - ControlNet - STATUS - ControlNet Hooked - Time = 19.470924139022827
100%|██████████████████████████████████████████████████████████████████████████████████| 15/15 [00:11<00:00,  1.31it/s]
2023-09-16 21:59:21,277 - ControlNet - STATUS - Loading model from cache: control_v11p_sd15_seg [e1f51eb9]
2023-09-16 21:59:33,935 - ControlNet - STATUS - Loading preprocessor: oneformer_ade20k
2023-09-16 21:59:33,935 - ControlNet - STATUS - preprocessor resolution = 600
2023-09-16 21:59:34,799 - ControlNet - STATUS - ControlNet Hooked - Time = 13.523083686828613
100%|██████████████████████████████████████████████████████████████████████████████████| 15/15 [00:11<00:00,  1.35it/s]
2023-09-16 22:00:35,732 - ControlNet - STATUS - Loading model from cache: control_v11p_sd15_seg [e1f51eb9]
2023-09-16 22:00:48,596 - ControlNet - STATUS - Loading preprocessor: oneformer_ade20k
2023-09-16 22:00:48,597 - ControlNet - STATUS - preprocessor resolution = 600
2023-09-16 22:00:48,629 - ControlNet - STATUS - ControlNet Hooked - Time = 12.896502256393433
100%|██████████████████████████████████████████████████████████████████████████████████| 15/15 [00:11<00:00,  1.36it/s]
2023-09-16 22:01:18,642 - ControlNet - STATUS - Loading model from cache: control_v11p_sd15_seg [e1f51eb9]
2023-09-16 22:01:31,378 - ControlNet - STATUS - Loading preprocessor: oneformer_ade20k
2023-09-16 22:01:31,378 - ControlNet - STATUS - preprocessor resolution = 600
2023-09-16 22:01:31,408 - ControlNet - STATUS - ControlNet Hooked - Time = 12.76684832572937
100%|██████████████████████████████████████████████████████████████████████████████████| 10/10 [00:07<00:00,  1.36it/s]
100%|██████████████████████████████████████████████████████████████████████████████████| 16/16 [00:08<00:00,  1.92it/s]
2023-09-16 22:02:35,455 - ControlNet - STATUS - Preview Resolution = 650
2023-09-16 22:02:35,490 - ControlNet - STATUS - Loaded state_dict from [E:\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\pidinet\table5_pidinet.pth]
2023-09-16 22:02:46,133 - ControlNet - STATUS - Loading model: control_v11p_sd15_softedge [a8575a2a]
2023-09-16 22:02:52,237 - ControlNet - STATUS - Loaded state_dict from [E:\stable-diffusion-webui\models\ControlNet\control_v11p_sd15_softedge.pth]
2023-09-16 22:02:52,237 - ControlNet - STATUS - controlnet_default_config
2023-09-16 22:02:54,530 - ControlNet - STATUS - ControlNet model control_v11p_sd15_softedge [a8575a2a] loaded.
2023-09-16 22:03:07,545 - ControlNet - STATUS - Loading preprocessor: pidinet
2023-09-16 22:03:07,545 - ControlNet - STATUS - preprocessor resolution = 600
2023-09-16 22:03:07,742 - ControlNet - STATUS - ControlNet Hooked - Time = 22.05817461013794
100%|██████████████████████████████████████████████████████████████████████████████████| 16/16 [00:12<00:00,  1.31it/s]
2023-09-16 22:03:54,092 - ControlNet - STATUS - Loading model from cache: control_v11p_sd15_softedge [a8575a2a]
2023-09-16 22:04:06,741 - ControlNet - STATUS - Loading preprocessor: pidinet
2023-09-16 22:04:06,741 - ControlNet - STATUS - preprocessor resolution = 600
2023-09-16 22:04:06,773 - ControlNet - STATUS - ControlNet Hooked - Time = 12.684131622314453
100%|██████████████████████████████████████████████████████████████████████████████████| 16/16 [00:11<00:00,  1.37it/s]
2023-09-16 22:05:01,544 - ControlNet - STATUS - Loading model from cache: control_v11p_sd15_softedge [a8575a2a]
2023-09-16 22:05:14,258 - ControlNet - STATUS - Loading preprocessor: pidinet
2023-09-16 22:05:14,258 - ControlNet - STATUS - preprocessor resolution = 600
2023-09-16 22:05:14,364 - ControlNet - STATUS - ControlNet Hooked - Time = 12.824077606201172
100%|██████████████████████████████████████████████████████████████████████████████████| 16/16 [00:11<00:00,  1.36it/s]
```

**Comment by MotherSoraka** [comment]
i'm facing the same issue, did you ever found a fix?

**Comment by XWalways** [comment]
same question, any solutions

**Comment by jiqizaisikao** [comment]
same question

**Comment by XWalways** [comment]
***@***.***

**Comment by huchenlei** [comment]
I cannot reproduce the issue you have. The `loading model from cache` log is normal.

I am not sure why your subsequent runs still take very long time, maybe you can try add `--controlnet-loglevel DEBUG` to your A1111 launch args.

Here are my debug logs. You can see that the first run cost 8s to hook while the second run takes <1s. Also, the preprocessor does not run again in the second run, as there is no ` Calling preprocessor openpose_full outside of cache.` log there.

```
2023-10-14 22:12:00,385 - ControlNet - INFO - Loading model: control_v11p_sd15_openpose [cab727d4]
2023-10-14 22:12:03,461 - ControlNet - INFO - Loaded state_dict from [D:\stable-diffusion-webui\extensions\sd-webui-controlnet\models\control_v11p_sd15_openpose.pth]
2023-10-14 22:12:03,462 - ControlNet - INFO - controlnet_default_config
2023-10-14 22:12:05,489 - ControlNet - INFO - ControlNet model control_v11p_sd15_openpose [cab727d4] loaded.
2023-10-14 22:12:05,613 - ControlNet - DEBUG - build_control_model ran in: 5.228952884674072 sec
2023-10-14 22:12:05,614 - ControlNet - DEBUG - load_control_model ran in: 5.229953765869141 sec
2023-10-14 22:12:05,614 - ControlNet - DEBUG - Safe numpy convertion START
2023-10-14 22:12:05,615 - ControlNet - DEBUG - Safe numpy convertion END
2023-10-14 22:12:05,615 - ControlNet - INFO - Loading preprocessor: openpose_full
2023-10-14 22:12:05,616 - ControlNet - INFO - preprocessor resolution = 512
2023-10-14 22:12:05,619 - ControlNet - DEBUG - Use numpy seed 463735650.
2023-10-14 22:12:05,620 - ControlNet - DEBUG - Calling preprocessor openpose_full outside of cache.
2023-10-14 22:12:08,045 - ControlNet - DEBUG - detectmap_proc ran in: 0.004034519195556641 sec
2023-10-14 22:12:08,131 - ControlNet - DEBUG - controlnet_main_entry ran in: 8.251023292541504 sec
2023-10-14 22:12:08,132 - ControlNet - INFO - ControlNet Hooked - Time = 8.25202488899231
2023-10-14 22:12:08,132 - ControlNet - DEBUG - controlnet_hack ran in: 8.25202488899231 sec
2023-10-14 22:12:08,132 - ControlNet - DEBUG - process ran in: 8.25202488899231 sec
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:03<00:00,  5.47it/s]
2023-10-14 22:12:12,383 - ControlNet - DEBUG - postprocess ran in: 0.26123785972595215 sec██████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:01<00:00,  9.92it/s]
Total progress: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:02<00:00,  8.30it/s] 
2023-10-14 22:12:33,257 - ControlNet - DEBUG - parse_remote_call ran in: 0.0010001659393310547 sec██████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:02<00:00,  9.92it/s] 
2023-10-14 22:12:33,257 - ControlNet - DEBUG - get_enabled_units ran in: 0.0010001659393310547 sec
2023-10-14 22:12:33,515 - ControlNet - INFO - Loading model from cache: control_v11p_sd15_openpose [cab727d4]
2023-10-14 22:12:33,516 - ControlNet - DEBUG - load_control_model ran in: 0.0010020732879638672 sec
2023-10-14 22:12:33,516 - ControlNet - DEBUG - Safe numpy convertion START
2023-10-14 22:12:33,517 - ControlNet - DEBUG - Safe numpy convertion END
2023-10-14 22:12:33,517 - ControlNet - INFO - Loading preprocessor: openpose_full
2023-10-14 22:12:33,517 - ControlNet - INFO - preprocessor resolution = 512
2023-10-14 22:12:33,517 - ControlNet - DEBUG - Use numpy seed 742064868.
2023-10-14 22:12:33,526 - ControlNet - DEBUG - detectmap_proc ran in: 0.004004478454589844 sec
2023-10-14 22:12:33,587 - ControlNet - DEBUG - controlnet_main_entry ran in: 0.33158302307128906 sec
2023-10-14 22:12:33,588 - ControlNet - INFO - ControlNet Hooked - Time = 0.3325841426849365
2023-10-14 22:12:33,588 - ControlNet - DEBUG - controlnet_hack ran in: 0.3325841426849365 sec
2023-10-14 22:12:33,588 - ControlNet - DEBUG - process ran in: 0.3325841426849365 sec
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:02<00:00,  8.83it/s]
2023-10-14 22:12:36,377 - ControlNet - DEBUG - postprocess ran in: 0.2552828788757324 sec███████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:02<00:00,  

> 9.52it/s]

```

**Comment by tothlevi** [comment]
Same problem here. Gets hooked quite slowly.

2023-11-08 13:12:57,030 - ControlNet - INFO - ControlNet model ip-adapter-plus-face_sd15 [71693645] loaded.
2023-11-08 13:12:57,044 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15
2023-11-08 13:12:57,044 - ControlNet - INFO - preprocessor resolution = 512
2023-11-08 13:13:04,837 - ControlNet - INFO - ControlNet Hooked - Time = 8.231678485870361
100%|##########| 16/16 [00:02<00:00,  6.02it/s]
Total progress: 100%|██████████| 16/16 [00:02<00:00,  5.52it/s]s]
2023-11-08 13:13:34,651 - AnimateDiff - INFO - Hacking i2i-batch.
2023-11-08 13:13:34,655 - AnimateDiff - INFO - Hacking i2i-batch.
2023-11-08 13:13:34,859 - ControlNet - INFO - Loading model from cache: ip-adapter-plus-face_sd15 [71693645]
2023-11-08 13:13:35,048 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15
2023-11-08 13:13:35,048 - ControlNet - INFO - preprocessor resolution = 512
2023-11-08 13:13:43,449 - ControlNet - INFO - ControlNet Hooked - Time = 8.71114468574524
100%|##########| 38/38 [00:06<00:00,  5.66it/s]
Total progress: 100%|██████████| 38/38 [00:05<00:00,  6.37it/s]s]
2023-11-08 13:16:08,065 - ControlNet - INFO - Loading model from cache: ip-adapter-plus-face_sd15 [71693645]
2023-11-08 13:16:08,208 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15
2023-11-08 13:16:08,208 - ControlNet - INFO - preprocessor resolution = 512
2023-11-08 13:16:16,767 - ControlNet - INFO - ControlNet Hooked - Time = 8.833007335662842
100%|##########| 16/16 [00:02<00:00,  6.67it/s]
Total progress: 100%|██████████| 16/16 [00:02<00:00,  6.14it/s]s]

**Comment by msha3er** [comment]
Does anyone fixed this issue?

**Comment by XWalways** [comment]
***@***.***

**Comment by vinch00** [comment]
Happening here as well. Here is the debug log.
Slow hook times are adding 15 to 20 seconds to each generation.

```
2024-01-24 02:30:49,543 - ControlNet - DEBUG - detectmap_proc ran in: 0.002 sec
2024-01-24 02:30:49,795 - ControlNet - INFO - Loading model: control_v11p_sd15_openpose [cab727d4]
2024-01-24 02:30:50,299 - ControlNet - INFO - Loaded state_dict from [E:\stable-diffusion-webui\models\ControlNet\control_v11p_sd15_openpose.pth]
2024-01-24 02:30:50,300 - ControlNet - INFO - controlnet_default_config
2024-01-24 02:30:52,322 - ControlNet - INFO - ControlNet model control_v11p_sd15_openpose [cab727d4] loaded.
2024-01-24 02:30:52,397 - ControlNet - DEBUG - build_control_model ran in: 2.602 sec
2024-01-24 02:30:52,398 - ControlNet - DEBUG - load_control_model ran in: 2.855 sec
2024-01-24 02:30:52,402 - ControlNet - DEBUG - choose_input_image ran in: 0.004 sec
2024-01-24 02:30:52,402 - ControlNet - DEBUG - is_upscale_script=False
2024-01-24 02:30:52,403 - ControlNet - DEBUG - Use numpy seed 403918.
2024-01-24 02:30:52,403 - ControlNet - INFO - Using preprocessor: openpose
2024-01-24 02:30:52,403 - ControlNet - INFO - preprocessor resolution = 512
2024-01-24 02:30:52,406 - ControlNet - DEBUG - detectmap_proc ran in: 0.003 sec
2024-01-24 02:30:52,675 - ControlNet - INFO - Loading model: control_v11p_sd15_canny [d14c016b]
2024-01-24 02:30:53,230 - ControlNet - INFO - Loaded state_dict from [E:\stable-diffusion-webui\models\ControlNet\control_v11p_sd15_canny.pth]
2024-01-24 02:30:53,231 - ControlNet - INFO - controlnet_default_config
2024-01-24 02:30:55,685 - ControlNet - INFO - ControlNet model control_v11p_sd15_canny [d14c016b] loaded.
2024-01-24 02:30:55,754 - ControlNet - DEBUG - build_control_model ran in: 3.079 sec
2024-01-24 02:30:55,754 - ControlNet - DEBUG - load_control_model ran in: 3.348 sec
2024-01-24 02:30:55,776 - ControlNet - DEBUG - choose_input_image ran in: 0.022 sec
2024-01-24 02:30:55,776 - ControlNet - DEBUG - is_upscale_script=False
2024-01-24 02:30:55,776 - ControlNet - DEBUG - Use numpy seed 403918.
2024-01-24 02:30:55,776 - ControlNet - INFO - Using preprocessor: canny
2024-01-24 02:30:55,776 - ControlNet - INFO - preprocessor resolution = 512
2024-01-24 02:30:55,781 - ControlNet - DEBUG - detectmap_proc ran in: 0.004 sec
2024-01-24 02:30:56,151 - ControlNet - INFO - Loading model: ip-adapter-plus_sd15 [836b5c2e]
2024-01-24 02:30:56,161 - ControlNet - INFO - Loaded state_dict from [E:\stable-diffusion-webui\models\ControlNet\ip-adapter-plus_sd15.safetensors]
2024-01-24 02:30:56,435 - ControlNet - INFO - ControlNet model ip-adapter-plus_sd15 [836b5c2e] loaded.
2024-01-24 02:30:56,436 - ControlNet - DEBUG - build_control_model ran in: 0.285 sec
2024-01-24 02:30:56,436 - ControlNet - DEBUG - load_control_model ran in: 0.654 sec
2024-01-24 02:30:56,440 - ControlNet - DEBUG - choose_input_image ran in: 0.003 sec
2024-01-24 02:30:56,441 - ControlNet - DEBUG - is_upscale_script=False
2024-01-24 02:30:56,441 - ControlNet - DEBUG - Use numpy seed 403918.
2024-01-24 02:30:56,442 - ControlNet - INFO - Using preprocessor: ip-adapter_clip_sd15
2024-01-24 02:30:56,442 - ControlNet - INFO - preprocessor resolution = 512
2024-01-24 02:30:56,443 - ControlNet - DEBUG - Calling preprocessor ip-adapter_clip_sd15 outside of cache.
2024-01-24 02:31:04,476 - ControlNet - DEBUG - choose_input_image ran in: 0.002 sec
2024-01-24 02:31:04,477 - ControlNet - DEBUG - is_upscale_script=False
2024-01-24 02:31:04,477 - ControlNet - DEBUG - Use numpy seed 403918.
2024-01-24 02:31:04,477 - ControlNet - INFO - Using preprocessor: reference_only
2024-01-24 02:31:04,477 - ControlNet - INFO - preprocessor resolution = 512
2024-01-24 02:31:04,478 - ControlNet - DEBUG - Calling preprocessor reference_only outside of cache.
2024-01-24 02:31:04,486 - ControlNet - DEBUG - detectmap_proc ran in: 0.008 sec
2024-01-24 02:31:04,818 - ControlNet - DEBUG - controlnet_main_entry ran in: 18.347 sec
2024-01-24 02:31:04,819 - ControlNet - INFO - ControlNet Hooked - Time = 18.34662675857544
2024-01-24 02:31:04,819 - ControlNet - DEBUG - controlnet_hack ran in: 18.348 sec
2024-01-24 02:31:04,819 - ControlNet - DEBUG - process ran in: 18.348 sec
```

**Comment by huchenlei** [comment]
Try increase model cache size in ControlNet settings. The default value is 1 so if you are using more than 1 models per generation, only the first model get cached.

**Comment by vinch00** [comment]
@huchenlei 

Thanks, model cache size did help to reduce hook times from 18sec to 7sec.
Seems all ControlNet models are caching now, except for IP Adapter.
Almost the entirety of the 7sec delay is when IP Adapter is loading.
Any ideas how to debug further?

```
2024-01-24 02:49:42,506 - ControlNet - INFO - Using preprocessor: ip-adapter_clip_sd15
2024-01-24 02:49:42,506 - ControlNet - INFO - preprocessor resolution = 512
2024-01-24 02:49:42,507 - ControlNet - DEBUG - Calling preprocessor ip-adapter_clip_sd15 outside of cache.
2024-01-24 02:49:49,302 - ControlNet - DEBUG - is_upscale_script=False
2024-01-24 02:49:49,302 - ControlNet - DEBUG - Use numpy seed 2424979904.
2024-01-24 02:49:49,302 - ControlNet - INFO - Using preprocessor: reference_only
2024-01-24 02:49:49,302 - ControlNet - INFO - preprocessor resolution = 512
2024-01-24 02:49:49,304 - ControlNet - DEBUG - Calling preprocessor reference_only outside of cache.
2024-01-24 02:49:49,313 - ControlNet - DEBUG - detectmap_proc ran in: 0.009 sec
2024-01-24 02:49:49,425 - ControlNet - DEBUG - controlnet_main_entry ran in: 7.227 sec
2024-01-24 02:49:49,425 - ControlNet - INFO - ControlNet Hooked - Time = 7.226851224899292
2024-01-24 02:49:49,425 - ControlNet - DEBUG - controlnet_hack ran in: 7.227 sec
2024-01-24 02:49:49,425 - ControlNet - DEBUG - process ran in: 7.227 sec
```

**Comment by huchenlei** [comment]
We explicitly don't cache ipadapter for now. I think the caching should be 1 layer lower. Now we are caching the plugable adapter object, which contains things that cannot be shared among units. We should cache the ipadapter model instead. However, this requires some major refacotring.
